src/porepy/compositional/materials.py

```diff
@@ -0,0 +1,476 @@
+    contain various constants and unit declarations in dicts, all in simple formats and
+    per se hashable.
+
+    The need for hashable material constants arises when fluid constants are used as a
+    base class for components and tracers, which are dynamically created and themselves
+    stored in standard dictionaries at various points in ``porepy.compositional``.
+
+    """
+
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+
+# The dataclass is frozen, since material parameters are not supposed to change, as well
+# as the units they are given in.
+# By using this in combination with keyword_only arguments for the construction of
+# materials, the user is forced to instantiate the constants with the right names of
+# constants, otherwise errors are raised (no separate checks required).
+# By providing default values to the fields, not every constant is required. The user
+# is expected to be aware which physics are used in the model.
+@dataclass(frozen=True, kw_only=True)
+class MaterialConstants:
+    """Material property container and conversion class.
+
+    The base clase identifies a material using a given :attr:`name`.
+    To define material properties of some kind, derive a dataclass with this class
+    as its base (as frozen and keywords-only data class).
+
+    Material constants are declared as fields (float or int) with default values.
+
+    Derived classes must have a class attribute :attr:`SI_units` containing information
+    about the the physical unit of each declared constant.
+
+    If the user wants the material to be presented in other than SI units, a ``units=``
+    kw-argument can be passed to declare the target, non-SI units (f.e. MPa instead of
+    Pa).
+
+    Important:
+        When instantiating a material constants data class, the constants must all
+        be strictly given in SI units. The conversion happens post-initialization for
+        the subsequent simulation.
+
+    Important:
+        Every derived class must have a class attribute :attr:`SI_units`, **annotated as
+        ClassVar**. This is to inform the base class about the used SI units.
+
+        For examples, see :class:`FluidConstants` or :class:`SolidConstants`.
+        For instructions on how to write composed units, see
+        :meth:`~porepy.models.units.Units.convert_units`.
+
+    """
+
+    # NOTE Annotating it as a ClassVar leads to the dataclasss decorator ignoring this
+    # in its machinery. The annotation must not be forgotten in derived classes.
+    SI_units: ClassVar[_HashableDict[str, str]] = _HashableDict()
+    """A dictionary containing the SI unit of every material constant defined by
+    a derived class.
+
+    E.g., ``{'pressure': 'Pa'}``
+
+    Every derived data class must have one implemented and annotated as a ``ClassVar``.
+
+    Note:
+        The ``ClassVar`` hints to ``dataclass`` that this field should not be part of
```

Comment:
The class attribute declares the SI units, e.g., for pressure `Pa`, contrary to the (non-SI) simulation units, they are universal in some sense.
The motivation was to have that declared in 1 spot in the code (programmatically accessible), instead of writing it in the docs *and* hardcoding it in the previously callable attributes.

Yes, it prevents you to have, say, two fluid constants where one has `Pa` and the other `MPa` or something else declared as the base unit.

Also, the unit conversion is done in a general manner for all attributes, from here declared SI units to simulation units.

I also wanted to make it more straight forward to add new constants to a class, without having to worry about conversion etc. Further development now needs essentially to declare simply the field, and add its unit to this attribute (like I did here for the critical pressure for example).